Password plugin, cPanel driver

julianmatz opened the pull request:
Add support for hash auth (cPanel Remote Access Key) in the Password plugin's cPanel driver and call the $xmlapi->listaccts() method to retrieve the cPanel account user that owns the domain (required for cPanel reseller accounts).

julianmatz committed:
Update config.inc.php.dist

Add support for cPanel remote access hash (choice between hash auth and plain password auth).

julianmatz committed:
Update config.inc.php.dist

Add support for cPanel hash auth, using a Remote Access Key (choice between hash auth and plain password auth).

julianmatz committed:
Update cpanel.php

Add support for hash auth (cPanel Remote Access Key) and call the $xmlapi->listaccts() method to retrieve the cPanel account user that owns the domain (required for cPanel reseller accounts).

alecpl commented:
Could you keep our code style and indentation, please?

julianmatz committed:
Update cpanel.php

Changed code style and indentation.

julianmatz committed:
Update cpanel.php

julianmatz commented:
Sure. I've fixed it in commit [8042955](https://github.com/roundcube/roundcubemail/pull/5252/commits/80429559271996f9ac7534c02ed8da29c24cb828)

alecpl commented on plugins/password/drivers/cpanel.php:
```diff
@@ -45,7 +45,18 @@ public function save($curpas, $newpass)
+        }
+        // Pass auth
+        else if (!empty($rcmail->config->get('password_cpanel_password'))) {
+            $this->xmlapi->hash_auth( $this->cuser, $rcmail->config->get('password_cpanel_password'));
```
Here password_hash() should be used. Also, I'd like to see config->get() use once per option.

julianmatz replied:
I've modified the code to call config->get() only once. Commit is here:
https://github.com/roundcube/roundcubemail/pull/5252/commits/7ed0d3308b2b5968cc739f3f1035c1b866661deb

The hash key here is a unique remote access key that's generated and stored in the WHM user's account. For example, for the root user, it would be stored in /root/.accesshash. I don't think it can be generated outside of the WHM interface.

https://documentation.cpanel.net/display/ALD/Remote+Access+Key

alecpl commented on plugins/password/drivers/cpanel.php:
```diff
@@ -45,7 +45,18 @@ public function save($curpas, $newpass)
+            $this->xmlapi->hash_auth( $this->cuser, $rcmail->config->get('password_cpanel_password'));
+        }
+        else {
+            return false;
```
Here, would be nice to return error code and message.

julianmatz replied:
I'm not too familiar with the error handling. Would it be done using the following? `$rcmail->output->command('display_message', 'message, 'error');`

alecpl commented on plugins/password/drivers/cpanel.php:
```diff
@@ -70,8 +81,16 @@ function setPassword($address, $password)
+        if ( $query['status'] != 1 ) {
+            return false;
+        }
+        $cpanel_user = $query['acct'][0]['user'];
```
We have username from config, why we need to overwrite it here? I don't know cPanel API, but can you call listaccts() when using user/password auth. Do we need it at all when not using hash auth?

julianmatz replied:
Yes, it's required whether the password or the hash is used. The username retrieved from config is the API user (root user or reseller user). The cPanel user needs to be retrieved using listaccts() because otherwise, the following error will be returned:

`You do not have an email account named “email.alias@example.com”.`

(Even if authenticated as root).

This can easily be tested on a cPanel server, using the following example URL:

`https://hostname.example.com:2087/cpsess##########/json-api/cpanel?cpanel_jsonapi_user=user&cpanel_jsonapi_apiversion=2&cpanel_jsonapi_module=Email&cpanel_jsonapi_func=passwdpop&domain="example.com"&email="user"&password="12345luggage"`

julianmatz committed:
Update cpanel.php

Call $rcmail->config->get() only once per option by assigning variable inside the IF statement.
